Pass bloom filter to SPV node

torkelrogstad opened the pull request:
In this PR we verify that passing a bloom filter from the wallet to the node actually triggers something with our Bitcoin Core peer. No further action is taken in the wallet upon receiving data.

torkelrogstad committed:
Add getAddressInfo to Wallet API

torkelrogstad committed:
Add Bloom filter and event callbacks to SPV node

torkelrogstad committed:
Move wallet test trait to testkit

torkelrogstad committed:
Test for connecting SPV node with wallet

Christewart commented on node-test/src/test/scala/org/bitcoins/node/NodeWithWalletTest.scala:
```diff
@@ -0,0 +1,143 @@
+                    bloomFilter = Some(bloom),
+                    callbacks = Vector(callbacks))
+          spv.start()
+        }
```
```scala
_ <- spv.sync()
```
is needed to actually start the sync

torkelrogstad committed:
Tune logging in SPV node

torkelrogstad committed:
Clean up BloomFilter

torkelrogstad committed:
Proper toString in inventory and inventorymessage

torkelrogstad committed:
Actually pass in callbacks to SPV node

nkohen commented on node-test/src/test/scala/org/bitcoins/node/NodeWithWalletTest.scala:
```diff
@@ -0,0 +1,136 @@
+
+  it must "load a bloom filter and receive information about received payments" in {
+    param =>
+      val WalletWithBitcoind(wallet, rpc) = param
```
any reason to not just do `case WalletWithBitcoind(wallet, rpc) => ...` ?

torkelrogstad replied:
That makes the compiler complain about expecting a function (as compared to a partial function) as the parameter to the test :-(

nkohen commented on node-test/src/test/scala/org/bitcoins/node/NodeWithWalletTest.scala:
```diff
@@ -0,0 +1,136 @@
+        * to resolve to multiple implicit parameters?
+        */
+      implicit val nodeConfig: NodeAppConfig = config
+      implicit val chainConfig: ChainAppConfig = config
```
Not sure this would work but perhaps `implicit val config: NodeAppConfig with ChainAppConfig`?

torkelrogstad replied:
Doesn't work :-(

nkohen commented on node-test/src/test/scala/org/bitcoins/node/NodeWithWalletTest.scala:
```diff
@@ -0,0 +1,136 @@
+            completionP.failure(
+              new TestFailedException(
+                s"Got ${tx.txId}, which is a TX we expect to NOT get notified about!",
+                failedCodeStackDepth = 0))
```
I think this should be
```
fail(s"Got ${tx.txId}, which is a TX we expect to NOT get notified about!")
```

torkelrogstad replied:
`fail` is similar to `???` in that when you try to evaluate it it throws: https://github.com/scalatest/scalatest/blob/441df125737543ce7e1776f087ca57e79ee17159/scalatest/src/main/scala/org/scalatest/Assertions.scala#L1101

So for the failure to propagate correctly I _think_ this is the way to do it

nkohen commented on node-test/src/test/scala/org/bitcoins/node/NodeWithWalletTest.scala:
```diff
@@ -0,0 +1,136 @@
+              val msg =
+                s"Did not receive sent transaction within $delay"
+              logger.error(msg)
+              completionP.failure(new TestFailedException(msg, 0))
```
same as above

nkohen commented on node/src/main/scala/org/bitcoins/node/SpvNode.scala:
```diff
@@ -15,16 +15,24 @@ import org.bitcoins.node.networking.peer.{
+    peer: Peer,
+    chainApi: ChainApi,
+    callbacks: Vector[SpvNodeCallbacks] = SpvNodeCallbacks.empty,
+    bloomFilter: Option[BloomFilter] = None
```
In what situation would this ever be `None`?

torkelrogstad replied:
You may want to sync with block filters, or maybe use this node to actually download everything from the network

But you're right that right now there's probably no use cases where this would be none. Can change it

Christewart replied:
Yeah, especially with the name of the class being `SpvNode` which implies bloom filters IMO.

nkohen commented on node/src/main/scala/org/bitcoins/node/networking/peer/DataMessageHandler.scala:
```diff
@@ -47,8 +68,31 @@ class DataMessageHandler()(
+  type OnTxReceived = Transaction => Unit
+
+  /** Does nothing with the received transaction */
+  val noopTxReceived: OnTxReceived = _ => ()
```
If you wanted you could just make a `def noop[T]: T => Unit = _ => ()` to avoid having all these vals?

torkelrogstad replied:
Good point

nkohen commented on node/src/main/scala/org/bitcoins/node/networking/peer/DataMessageHandler.scala:
```diff
@@ -5,28 +5,40 @@ import org.bitcoins.chain.blockchain.ChainHandler
   * [[HeadersMessage]] we should store those headers in our database
   */
-class DataMessageHandler()(
+class DataMessageHandler(callbacks: Vector[SpvNodeCallbacks])(
```
Rather than using `Vector[SpvNodeCallbacks]` everywhere which requires we have the same number of each kind of callback specified (resulting in noops), is there any reason we don't instead let `SpvNodeCallbacks` have three `Vector`s instead of three vals?

nkohen replied:
The abstraction just seems like an odd module to me

nkohen commented on node/src/main/scala/org/bitcoins/node/networking/peer/DataMessageHandler.scala:
```diff
@@ -38,6 +50,15 @@ class DataMessageHandler()(
+        FutureUtil.unit
+      case msg: MerkleBlockMessage =>
+        callbacks.foreach(_.onMerkleBlockReceived(msg.merkleBlock))
+        FutureUtil.unit
```
If you wrap the callback lines in `Future` then failures will nicely result in failed `Future`s rather than `throw`s

nkohen commented on testkit/src/main/scala/org/bitcoins/testkit/node/NodeTestUtil.scala:
```diff
@@ -95,6 +99,27 @@ abstract class NodeTestUtil {
+  def awaitSync(node: SpvNode, rpc: BitcoindRpcClient)(
+      implicit sys: ActorSystem): Future[Unit] = {
+    import sys.dispatcher
+    RpcUtil
```
I think that using `AsyncTestUtil` rather than `RpcUtil` results in nice `TestFailedException`s on failure

nkohen commented on wallet/src/main/scala/org/bitcoins/wallet/LockedWallet.scala:
```diff
@@ -314,6 +314,19 @@ abstract class LockedWallet extends LockedWalletApi with BitcoinSLogger {
   }
 
+  override def getAddressInfo(
+      address: BitcoinAddress): Future[Option[AddressInfo]] = {
```
Thoughts on flattening this to just a `Future`? How explicitly do you expect users to need to deal with the `None` case very differently from the `Future.failed` case?

torkelrogstad replied:
My reasoning here is twofold: 

1) A failed future isn't encoded into the type, which is bad. In that way its similar to `Option.get`
2) A failed future might be triggered by a DB problem, a network problem, a file system problem, or a bunch of things. Whereas a `None` cleanly indicates that the address isn't in the wallet

Christewart replied:
I do agree with this reasoning, but it does make the api harder to work with with `for` expressions

nkohen replied:
I would argue that `Future[T]` is just a wrapper for `Option[Try[T]]` but that point aside, I think that if the future is failed with a descriptive error then the issue of distinguishing a missing address from a db problem isn't an issue

torkelrogstad replied:
Use case for having this return `Future[Option[..]]`: 

A user wants to detect if an address is _not_ in the wallet. With your approach the user would then either match on the type of failure in the future (which he gets no help from the compiler doing, as futures aren't parameterized on the error type), or check if the error message on the exception matches a certain substring (which is bad)

nkohen replied:
I would say there's calls on Futures like `transform` and `onComplete` and `recover` that would let you handle this nicely if the error used in failure has its own type specific to a missing address but at the same time I don't think this is a huge deal so if you still think that `Future[Option]` is the way to go I'm okay with going with it and having the discussion if need be when it gets used in the future

Christewart commented on node-test/src/test/scala/org/bitcoins/node/NodeWithWalletTest.scala:
```diff
@@ -0,0 +1,136 @@
+        _ <- spv.sync()
+        _ <- NodeTestUtil.awaitSync(spv, rpc)
+
+        ourTxid <- rpc.sendToAddress(address, 1.bitcoin).map { tx =>
```
Would prefer to break this `rpc.sendToAddress` lambda into a separate method.

Christewart commented on core/src/main/scala/org/bitcoins/core/bloom/BloomFilter.scala:
```diff
@@ -142,16 +142,18 @@ sealed abstract class BloomFilter extends NetworkElement {
-        val constants = scriptPubKey.asm.filter(_.isInstanceOf[ScriptConstant])
-        constants.map(c => (c, index))
+        scriptPubKey.asm.collect {
+          case c: ScriptConstant => (c, index)
```
Just a heads up, I think i have a bug here implementing this. This won't be _only_ data pushes, but will include things like `ScriptNumber`, `OP_0`,`OP_1`, etc...

Christewart replied:
There are quite a few things that inherit from `ScriptConstant`

torkelrogstad committed:
Fix rebase screwup and partially respond to code review

torkelrogstad committed:
Default to file based databases in tests

This is due to issues with deadlocks with
in-memory based databases. We need DBs
to stay alive between connections, but
not across tests.

In DB intensive chain validation tests we
use in-memory databases instead. This
seems like a reasonable tradeoff between
simplicity and speed.

torkelrogstad committed:
Make SpvNodeCallbacks contain sequence of functions within

torkelrogstad committed:
Make Bloom filter non-optional in SpvNode
